### DISCUSSION TITLE
Sortが発生する場合とその対策_WHERE句でかける条件は、HAVING句に書かない方が効率がいい理由

### DISCUSSION THREAD

**Discussion by yuyakawai0411** [kakeru-one/web-engineering-roundtable#64] [General]
## 背景
こちらでWHERE句でかける条件は、HAVING句に書かない方が効率がいい理由を挙げているが、もう1つあると思った。
https://github.com/ochi-sho-private-study/mysql-sandbox/blob/main/docs/the_complete_guide_to_sql/sort.md

## 実行計画
```sql
EXPLAIN SELECT sale_data, SUM(quantity) FROM sales_history GROUP BY sale_data HAVING sale_data='2024-10-11'
```

| id | select_type | table         | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra           |
| --- | --- | ---         | --- | --- | --- | ---  | --- | ---  | --- | --- | --- |
|  1 | SIMPLE      | sales_history | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    7 |   100.00 | Using temporary |

```sql
EXPLAIN SELECT sale_data, SUM(quantity)  FROM sales_history WHERE sale_data = '2007-10-11';GROUP BY sale_data
```

| id | select_type | table         | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
| --- | --- | ---         | --- | --- | --- | ---  | --- | ---  | --- | --- | --- |
|  1 | SIMPLE      | sales_history | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    7 |    14.29 | Using where |

## 結論
- HAVING句を使うと、一時テーブルにテーブル情報を保存するためメモリを圧迫するから良くない。

**Comment by kakeru-one** [comment]
## SQLの実行順序
![スクリーンショット 2024-07-14 15 15 58](https://github.com/user-attachments/assets/7caeb044-859d-42c2-b43a-d03d693cb2c1)

### 参考文献
- https://speakerdeck.com/soudai/pgcon21j-tutorial?slide=15
